lib/src/merged_tree.rs

```diff
@@ -504,3 +518,200 @@ impl Iterator for ConflictIterator {
         }
     }
 }
+
+// TODO: Much of this code can probably be shared with
+// `tree::TreeEntryDiffIterator` by adding some traits with associated types.
+struct TreeEntryDiffIterator<'a> {
+    before: &'a MergedTree,
+    after: &'a MergedTree,
+    basename_iter: Box<dyn Iterator<Item = &'a RepoPathComponent> + 'a>,
+}
+
+impl<'a> TreeEntryDiffIterator<'a> {
+    fn new(before: &'a MergedTree, after: &'a MergedTree) -> Self {
+        fn merge_iters<'a>(
+            iter1: impl Iterator<Item = &'a RepoPathComponent> + 'a,
+            iter2: impl Iterator<Item = &'a RepoPathComponent> + 'a,
+        ) -> Box<dyn Iterator<Item = &'a RepoPathComponent> + 'a> {
+            Box::new(iter1.merge(iter2).dedup())
+        }
+        let basename_iter: Box<dyn Iterator<Item = &'a RepoPathComponent> + 'a> =
+            match (before, after) {
+                (MergedTree::Legacy(before), MergedTree::Legacy(after)) => {
+                    merge_iters(before.data().names(), after.data().names())
+                }
+                (MergedTree::Merge(before), MergedTree::Legacy(after)) => {
+                    merge_iters(all_tree_conflict_names(before), after.data().names())
+                }
+                (MergedTree::Legacy(before), MergedTree::Merge(after)) => {
+                    merge_iters(before.data().names(), all_tree_conflict_names(after))
+                }
+                (MergedTree::Merge(before), MergedTree::Merge(after)) => merge_iters(
+                    all_tree_conflict_names(before),
+                    all_tree_conflict_names(after),
+                ),
+            };
+        TreeEntryDiffIterator {
+            before,
+            after,
+            basename_iter,
+        }
+    }
+}
+
+impl<'a> Iterator for TreeEntryDiffIterator<'a> {
+    type Item = (
+        &'a RepoPathComponent,
+        MergedTreeValue<'a>,
+        MergedTreeValue<'a>,
+    );
+
+    fn next(&mut self) -> Option<Self::Item> {
+        for basename in self.basename_iter.by_ref() {
+            let value_before = self.before.value(basename);
+            let value_after = self.after.value(basename);
+            if value_after != value_before {
+                return Some((basename, value_before, self.after.value(basename)));
```

Comment:
nit: `value_after` instead of `self.after.value...`

Reply:
Thanks for catching that.